torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,423 @@
+        tensor_name (str): name of the parameter, buffer, or parametrization
+            from which the parametrization will be applied
+    """
+    if not is_parametrized(module, tensor_name):
+        raise ValueError(
+            "The tensor {} in module {} is not parametrized".format(
+                tensor_name, type(module)
+            )
+        )
+
+    key = _key(module, tensor_name)
+    if key not in _cache:
+        _cache[key] = None
+
+
+def remove_caching(module: Module, tensor_name: str) -> None:
+    r"""Removes the caching mechanism for a given parametrization.
+    After calling this function, the values of the parametrization
+    will not be cached even context manager :func:`cached` is active.
+    This function is the inverse of :func:`set_caching`.
+
+    Args:
+        module (nn.Module): module on which to remove the caching mechanism
+        tensor_name (str): name of the parameter, buffer, or parametrization
+            from which the parametrization will be applied
+    """
+    if not is_parametrized(module, tensor_name):
+        raise ValueError(
+            "The tensor {} in module {} is not parametrized".format(
+                tensor_name, type(module)
+            )
+        )
+
+    key = _key(module, tensor_name)
+    if key in _cache:
+        _cache.pop(key)
+
+
+def has_caching(module: Module, tensor_name: str) -> bool:
+    r"""Returns ``True`` if ``module[tensor_name]`` is parametrized and
+    its values are cached when :func:`cached` is active.
+
+    Args:
+        module (nn.Module): module to query
+        tensor_name (str): attribute in the module to query
+    """
+    key = _key(module, tensor_name)
+    return is_parametrized(module, tensor_name) and key in _cache
+
+
+def _inject_new_class(module: Module) -> None:
+    r"""Sets up the parametrization mechanism used by parametrizations.
+    This works by substituting the class of the module by a class
+    that extends it to be able to inject a property
+
+    Args:
+        module (nn.Module): module into which to inject the property
+    """
+    cls = module.__class__
+
+    def getstate(self):
+        raise RuntimeError(
+            "Serialization of parametrized modules is only "
+            "supported through state_dict(). See:\n"
+            "https://pytorch.org/tutorials/beginner/saving_loading_models.html"
+            "#saving-loading-a-general-checkpoint-for-inference-and-or-resuming-training"
+        )
+
+    param_cls = type(
+        "Parametrized{}".format(cls.__name__),
+        (cls,),
+        {
+            "__getstate__": getstate,
+        },
+    )
+
+    module.__class__ = param_cls
+
+
+def _inject_property(module: Module, tensor_name: str) -> None:
+    r"""Injects a property into module[tensor_name].
+    It assumes that the class in the module has already been modified from its
+    original one using _inject_new_class and that the tensor under `tensor_name`
+    has already been moved out
```

Comment:
This is already done in `register_parametrization`, as we do a `delattr(module, tensor_name`. It can be thought as a precondition to this function. That's the reason why I wrote that "it assumes that the tensor has already been moved out"

Reply:
But if you can check your assumptions easily, it is always better to do so